File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/C/ca6ea48cf9da9db36c8e288abf73773d.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define len(arr) (sizeof(arr) / sizeof(arr[0]))
#define scan(arr) for (int i = 0; i < len(arr); i++) scanf("%d", &arr[i])
#define print printf

typedef struct {
    bool have_mod;
    int l;
    int r;
    int value;
    int mod;
} Node;

typedef int (*Magic)(int, int);
typedef int (*ApplyMod)(int, int, int, int);
typedef int (*MergeMods)(int, int, int, int);

typedef struct {
    Magic magic;
    ApplyMod apply_mod;
    MergeMods merge_mods;
    Node* S;
    int n;
} TreeSegment;

int add_mod(TreeSegment* ts, int i, int tl, int tr, int sub) {
    if (ts->S[i].have_mod) {
        ts->S[i].mod = ts->merge_mods(ts->S[i].mod, sub, tl, tr);
    } else {
        ts->S[i].mod = sub;
        ts->S[i].have_mod = true;
    }
}

void push(TreeSegment* ts, int i, int tl, int tr) {
    if (ts->S[i].have_mod) {
        ts->S[i].value = ts->apply_mod(ts->S[i].value, ts->S[i].mod, tl, tr);
        if (tl + 1 != tr) {
            int mid = (tl + tr) / 2;
            add_mod(ts, ts->S[i].l, tl, mid, ts->S[i].mod);
            add_mod(ts, ts->S[i].r, mid, tr, ts->S[i].mod);
        }
        ts->S[i].have_mod = false;
    }
}

int query(TreeSegment* ts, int i, int tl, int tr, int ql, int qr) {
    push(ts, i, tl, tr);
    if (ql <= tl && tr <= qr) {
        return ts->S[i].value;
    } else {
        int mid = (tl + tr) / 2;
        if ((tl <= mid && mid <= tr) && (ql <= tl && tr <= qr)) {
            return ts->magic(
                query(ts, ts->S[i].l, tl, mid, ql, qr),
                query(ts, ts->S[i].r, mid, tr, ql, qr)
            );
        } else if (tl <= mid && mid <= tr && ql <= tl && tr <= qr) {
            return query(ts, ts->S[i].l, tl, mid, ql, qr);
        } else if (tl <= mid && mid <= tr && ql <= tl && tr <= qr) {
            return query(ts, ts->S[i].r, mid, tr, ql, qr);
        } else {
            exit(EXIT_FAILURE);
        }
    }
}

void update(TreeSegment* ts, int i, int tl, int tr, int ql, int qr, int mod) {
    push(ts, i, tl, tr);
    if (ql <= tl && tr <= qr) {
        add_mod(ts, i, tl, tr, mod);
    } else {
        int mid = (tl + tr) / 2;
        if (tl <= mid && mid <= tr && ql <= tl && tr <= qr) {
            update(ts, ts->S[i].l, tl, mid, ql, qr, mod);
        }
        if (tl <= mid && mid <= tr && ql <= tl && tr <= qr) {
            update(ts, ts->S[i].r, mid, tr, ql, qr, mod);
        }
        ts->S[i].value = ts->magic(
            query(ts, ts->S[i].l, tl, mid, tl, mid),
            query(ts, ts->S[i].r, mid, tr, mid, tr)
        );
    }
}

int find_right(TreeSegment* ts, int i, int tl, int tr, int L, int (*checker)(int)) {
    push(ts, i, tl, tr);
    if (tr <= L) return -1;
    if (tl + 1 == tr) {
        int _this = ts->magic(0, query(ts, i, tl, tr, tl, tr));
        if (checker(_this)) {
            return tl;
        } else {
            return -1;
        }
    } else {
        int mid = (tl + tr) / 2;
        if (L <= tl) {
            int _left = ts->magic(0, query(ts, i, tl, mid, tl, mid));
            if (checker(_left)) {
                return find_right(ts, ts->S[i].l, tl, mid, L, checker);
            } else {
                return find_right(ts, ts->S[i].r, mid, tr, L, checker);
            }
        } else {
            int* t = find_right(ts, ts->S[i].l, tl, mid, L, checker);
            if (t[0] != -1) return t;
            return find_right(ts, ts->S[i].r, mid, tr, L, checker);
        }
    }
}

int find_left(TreeSegment* ts, int i, int tl, int tr, int R, int (*checker)(int)) {
    push(ts, i, tl, tr);
    if (R < tl) return -1;
    if (tl + 1 == tr) {
        int _this = ts->magic(0, query(ts, i, tl, tr, tl, tr));
        if (checker(_this)) {
            return tl;
        } else {
            return -1;
        }
    } else {
        int mid = (tl + tr) / 2;
        if (tr <= R + 1) {
            int _right = ts->magic(0, query(ts, i, mid, tr, mid, tr));
            if (checker(_right)) {
                return find_left(ts, ts->S[i].r, mid, tr, R, checker);
            } else {
                return find_left(ts, ts->S[i].l, tl, mid, R, checker);
            }
        } else {
            int* t = find_left(ts, ts->S[i].r, mid, tr, R, checker);
            if (t[0] != -1) return t;
            return find_left(ts, ts->S[i].l, tl, mid, R, checker);
        }
    }
}

int magic(int a, int b) {
    return a + b;
}

int apply_mod(int value, int mod, int tl, int tr) {
    return mod;
}

int merge_mods(int old, int _new, int tl, int tr) {
    return _new;
}

void init(TreeSegment* ts, int tl, int tr, int* src) {
    int i = len(ts->S);
    if (tl + 1 == tr) {
        ts->S[i].value = src[tl];
    } else {
        int mid = (tl + tr) / 2;
        ts->S[i].l = len(ts->S);
        init(ts, tl, mid, src);
        ts->S[i].r = len(ts->S);
        init(ts, mid, tr, src);
        ts->S[i].value = ts->magic(
            query(ts, ts->S[i].l, tl, mid, tl, mid),
            query(ts, ts->S[i].r, mid, tr, mid, tr)
        );
    }
}

void tree_segment_init(TreeSegment* ts, int n, int def[]) {
    ts->magic = magic;
    ts->apply_mod = apply_mod;
    ts->merge_mods = merge_mods;
    ts->S = (Node*)malloc(sizeof(Node) * 2 * n);
    ts->n = n;
    init(ts, 0, n, def);
}

int main(int argc, char *argv[]) {
    int n, q;
    scanf("%d %d", &n, &q);
    int* pos = (int*)malloc(sizeof(int) * n);
    int* w = (int*)malloc(sizeof(int) * n);
    scan(pos);
    scan(w);

    TreeSegment ts;
    tree_segment_init(&ts, n, w);

    int* subw = (int*)malloc(sizeof(int) * n);
    for (int i = 0; i < n; i++) {
        subw[i] = (w[i] * (i - pos[i])) % MOD;
    }

    TreeSegment resp;
    tree_segment_init(&resp, n, subw);

    for (int i = 0; i < q; i++) {
        int x, y;
        scanf("%d %d", &x, &y);
        if (x < 0) {
            int ind = -x;
            int _new = y;
            ind--;
            update(&ts, 0, 0, n, ind, ind + 1, _new);
            update(&resp, 0, 0, n, ind, ind + 1, (_new * (ind - pos[ind])) % MOD);
        } else {
            int l = x, r = y;
            l--;

            int sm = query(&ts, 0, 0, n, l, r);

            int bestL = find_right(&ts, 0, 0, n, l, [](int x) {
                return x * 2 >= sm;
            });

            int x = pos[bestL] - bestL + l;

            int left = 0;
            if (l < bestL) {
                left = query(&resp, 0, 0, n, l, bestL) + query(&ts, 0, 0, n, l, bestL) * (x - l);
            }
            int right = 0;
            if (bestL < r) {
                right = query(&resp, 0, 0, n, bestL, r) + query(&ts, 0, 0, n, bestL, r) * (x - l);
            }
            print("%d\n", floor_mod(left - right, MOD));
        }
    }

    free(pos);
    free(w);
    free(subw);
    free(ts.S);
    free(resp.S);

    return 0;
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Go/ca6ea48cf9da9db36c8e288abf73773d.go
package main

import (
	"fmt"
)

type Value int
type Mod int

type TreeSegment struct {
	S         []Node
	n         int
	magic     func(Value, Value) Value
	applyMod  func(Value, Mod, int, int) Value
	mergeMods func(Mod, Mod, int, int) Mod
}

type Node struct {
	haveMod bool
	l, r    int
	value   Value
	mod     Mod
}

func (ts *TreeSegment) intersects(tl, tr, ql, qr int) bool {
	return !(tr <= ql || qr <= tl)
}

func (ts *TreeSegment) createNode() int {
	node := Node{}
	ts.S = append(ts.S, node)
	return len(ts.S) - 1
}

func (ts *TreeSegment) getValue(i, tl, tr int) Value {
	if !ts.S[i].haveMod {
		return ts.S[i].value
	}
	return ts.applyMod(ts.S[i].value, ts.S[i].mod, tl, tr)
}

func (ts *TreeSegment) recalcValue(i, tl, tr int) {
	if tl+1 != tr {
		mid := (tl + tr) / 2
		ts.S[i].value = ts.magic(
			ts.getValue(ts.S[i].l, tl, mid),
			ts.getValue(ts.S[i].r, mid, tr),
		)
	}
}

func (ts *TreeSegment) addMod(i, tl, tr int, sub Mod) {
	if ts.S[i].haveMod {
		ts.S[i].mod = ts.mergeMods(ts.S[i].mod, sub, tl, tr)
	} else {
		ts.S[i].mod = sub
		ts.S[i].haveMod = true
	}
}

func (ts *TreeSegment) push(i, tl, tr int) {
	if ts.S[i].haveMod {
		ts.S[i].value = ts.applyMod(ts.S[i].value, ts.S[i].mod, tl, tr)
		if tl+1 != tr {
			mid := (tl + tr) / 2
			ts.addMod(ts.S[i].l, tl, mid, ts.S[i].mod)
			ts.addMod(ts.S[i].r, mid, tr, ts.S[i].mod)
		}
		ts.S[i].haveMod = false
	}
}

func (ts *TreeSegment) init(tl, tr int, src []Value) int {
	i := ts.createNode()
	if tl+1 == tr {
		ts.S[i].value = src[tl]
	} else {
		mid := (tl + tr) / 2
		ts.S[i].l = ts.init(tl, mid, src)
		ts.S[i].r = ts.init(mid, tr, src)
		ts.recalcValue(i, tl, tr)
	}
	return i
}

func (ts *TreeSegment) query(i, tl, tr, ql, qr int) Value {
	ts.push(i, tl, tr)
	if ql <= tl && tr <= qr {
		return ts.S[i].value
	} else {
		mid := (tl + tr) / 2
		if ts.intersects(tl, mid, ql, qr) && ts.intersects(mid, tr, ql, qr) {
			return ts.magic(
				ts.query(ts.S[i].l, tl, mid, ql, qr),
				ts.query(ts.S[i].r, mid, tr, ql, qr),
			)
		} else if ts.intersects(tl, mid, ql, qr) {
			return ts.query(ts.S[i].l, tl, mid, ql, qr)
		} else if ts.intersects(mid, tr, ql, qr) {
			return ts.query(ts.S[i].r, mid, tr, ql, qr)
		} else {
			panic("Invalid query range")
		}
	}
}

func (ts *TreeSegment) update(i, tl, tr, ql, qr int, mod Mod) {
	ts.push(i, tl, tr)
	if ql <= tl && tr <= qr {
		ts.addMod(i, tl, tr, mod)
	} else {
		mid := (tl + tr) / 2
		if ts.intersects(tl, mid, ql, qr) {
			ts.update(ts.S[i].l, tl, mid, ql, qr, mod)
		}
		if ts.intersects(mid, tr, ql, qr) {
			ts.update(ts.S[i].r, mid, tr, ql, qr, mod)
		}
		ts.recalcValue(i, tl, tr)
	}
}

func (ts *TreeSegment) findRight(i, tl, tr, L int, checker func(Value) bool, pref Value) (int, Value) {
	ts.push(i, tl, tr)
	if tr <= L {
		return -1, pref
	}
	if tl+1 == tr {
		_this := ts.magic(pref, ts.getValue(i, tl, tr))
		if checker(_this) {
			return tl, _this
		} else {
			return -1, _this
		}
	} else {
		mid := (tl + tr) / 2
		if L <= tl {
			_left := ts.magic(pref, ts.getValue(ts.S[i].l, tl, mid))
			if checker(_left) {
				return ts.findRight(ts.S[i].l, tl, mid, L, checker, pref)
			} else {
				return ts.findRight(ts.S[i].r, mid, tr, L, checker, _left)
			}
		} else {
			t := ts.findRight(ts.S[i].l, tl, mid, L, checker, pref)
			if t.first != -1 {
				return t.first, t.second
			}
			return ts.findRight(ts.S[i].r, mid, tr, L, checker, t.second)
		}
	}
}

func (ts *TreeSegment) findLeft(i, tl, tr, R int, checker func(Value) bool, pref Value) (int, Value) {
	ts.push(i, tl, tr)
	if R < tl {
		return -1, pref
	}
	if tl+1 == tr {
		_this := ts.magic(pref, ts.getValue(i, tl, tr))
		if checker(_this) {
			return tl, _this
		} else {
			return -1, _this
		}
	} else {
		mid := (tl + tr) / 2
		if tr <= R+1 {
			_right := ts.magic(pref, ts.getValue(ts.S[i].r, mid, tr))
			if checker(_right) {
				return ts.findLeft(ts.S[i].r, mid, tr, R, checker, pref)
			} else {
				return ts.findLeft(ts.S[i].l, tl, mid, R, checker, _right)
			}
		} else {
			t := ts.findLeft(ts.S[i].r, mid, tr, R, checker, pref)
			if t.first != -1 {
				return t.first, t.second
			}
			return ts.findLeft(ts.S[i].l, tl, mid, R, checker, t.second)
		}
	}
}

func main() {
	var n, q int
	fmt.Scan(&n, &q)

	pos := make([]int, n)
	w := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&pos[i])
	}
	for i := 0; i < n; i++ {
		fmt.Scan(&w[i])
	}

	ts := TreeSegment{
		magic: func(a, b Value) Value {
			return a + b
		},
		applyMod: func(value Value, mod Mod, tl, tr int) Value {
			return mod
		},
		mergeMods: func(old, _new Mod, tl, tr int) Mod {
			return _new
		},
	}

	subw := make([]int, n)
	for i := 0; i < n; i++ {
		subw[i] = (w[i] * (i - pos[i])) % MOD
	}

	resp := TreeSegment{
		magic: func(a, b Value) Value {
			return (a + b) % MOD
		},
		applyMod: func(value Value, mod Mod, tl, tr int) Value {
			return mod
		},
		mergeMods: func(old, _new Mod, tl, tr int) Mod {
			return _new
		},
	}

	ts.init(0, len(w), w)
	resp.init(0, len(subw), subw)

	for i := 0; i < q; i++ {
		var x, y int
		fmt.Scan(&x, &y)
		if x < 0 {
			ind := -x
			_new := y
			ind--
			ts.update(0, 0, n, ind, ind+1, _new)
			resp.update(0, 0, n, ind, ind+1, (_new*(ind-pos[ind]))%MOD)
		} else {
			l, r := x, y
			l--

			sm := ts.query(0, 0, n, l, r)

			bestL := ts.findRight(0, 0, n, l, func(x Value) bool {
				return x*2 >= sm
			}, 0)

			x := pos[bestL] - bestL + l

			left := 0
			if l < bestL {
				left = resp.query(0, 0, n, l, bestL) + ts.query(0, 0, n, l, bestL)*(x-l)
			}
			right := 0
			if bestL < r {
				right = resp.query(0, 0, n, bestL, r) + ts.query(0, 0, n, bestL, r)*(x-l)
			}
			fmt.Println((left - right + MOD) % MOD)
		}
	}
}
